crates/control/src/calibration_controller.rs

```diff
@@ -0,0 +1,277 @@
+    measurement_bottom: PerceptionInput<
+        Option<CalibrationCaptureResponse<Measurement>>,
+        "VisionBottom",
+        "calibration_measurement?",
+    >,
+    measurement_top: PerceptionInput<
+        Option<CalibrationCaptureResponse<Measurement>>,
+        "VisionTop",
+        "calibration_measurement?",
+    >,
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    look_at_dispatch_delay: Parameter<Duration, "calibration_controller.look_at_dispatch_delay">,
+    stabilization_delay: Parameter<Duration, "calibration_controller.stabilization_delay">,
+    max_retries_per_capture: Parameter<u32, "calibration_controller.max_retries_per_capture">,
+
+    calibration_measurements: AdditionalOutput<Vec<Measurement>, "calibration_measurements">,
+    last_calibration_corrections:
+        AdditionalOutput<Option<Corrections>, "last_calibration_corrections">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub calibration_command: MainOutput<Option<CalibrationCommand>>,
+}
+
+#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize)]
+enum CalibrationState {
+    #[default]
+    Inactive,
+    Initialize {
+        started_time: CycleTime,
+    },
+    LookAt {
+        target: Point2<Ground>,
+        camera: CameraPosition,
+        dispatch_time: CycleTime,
+    },
+    Capture {
+        camera: CameraPosition,
+        dispatch_time: CycleTime,
+    },
+    Finish,
+}
+
+impl CalibrationState {
+    fn as_calibration_command(&self) -> Option<CalibrationCommand> {
+        match *self {
+            CalibrationState::LookAt {
+                target,
+                camera,
+                dispatch_time,
+            } => Some(CalibrationCommand::LookAt {
+                target,
+                camera,
+                dispatch_time,
+            }),
+            CalibrationState::Capture {
+                camera,
+                dispatch_time,
+            } => Some(CalibrationCommand::Capture {
+                camera,
+                dispatch_time,
+            }),
+            _ => None,
+        }
+    }
+}
+
+impl CalibrationController {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            current_calibration_state: CalibrationState::default(),
+            look_at_list: generate_look_at_list(),
+            current_look_at_index: 0,
+            current_measurements: vec![],
+            last_calibration_corrections: None,
+            last_capture_retries: 0,
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        if *context.primary_state != PrimaryState::Calibration {
+            self.current_calibration_state = CalibrationState::Inactive;
+            return Ok(MainOutputs::default());
+        }
+
+        match self.current_calibration_state {
+            CalibrationState::Inactive => {
+                self.current_measurements = vec![];
```

Comment:
You should also reset the current_look_at_index here.
Currently, when trying to calibrate twice, the measurements are cleared, but since there are no more targets left, you immediately transition from Initialize to Finish and end up with no measurements at all.

Reply:
Oh damn! Good catch! Should define a reset function or a seperate struct with the mutable parts so re-start means just create it freshly

Reply:
Having a separate struct that can just be reassigned is probably the cleanest approach, yes.

Reply:
Did it, looks a little odd, but I think it isn't too bad:
```rust

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
struct StateTracking {
    look_at_index: usize,
    calibration_state: CalibrationState,
    measurements: Vec<Measurement>,
    last_capture_retries: u32,
}
```